File: Abstract Factory/Exercicios/Ex1_Transporte/Ex1.js
/**
 * PADRÃO: ABSTRACT FACTORY (FÁBRICA ABSTRATA)
 * 
 * Exercício: Sistema de Transporte.
 * O padrão garante que veículos individuais e coletivos pertençam ao mesmo 
 * ambiente (Terrestre ou Aéreo) de forma consistente.
 */

// 1. PRODUTOS ABSTRATOS
class VeiculoIndividual {
    iniciarRota() { throw new Error("Método abstrato"); }
}

class VeiculoColetivo {
    iniciarRota() { throw new Error("Método abstrato"); }
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA TERRESTRE
class Carro extends VeiculoIndividual {
    iniciarRota() { return "Partiu! Carro na pista (Transporte Terrestre Individual)."; }
}

class Onibus extends VeiculoColetivo {
    iniciarRota() { return "Lotou! Ônibus saindo do ponto (Transporte Terrestre Coletivo)."; }
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA AÉREA
class Helicoptero extends VeiculoIndividual {
    iniciarRota() { return "Hélice girando. Decolagem autorizada! (Transporte Aéreo Individual)."; }
}

class Aviao extends VeiculoColetivo {
    iniciarRota() { return "Atenção passageiros, iniciando voo (Transporte Aéreo Coletivo)."; }
}

// 3. FABRICA ABSTRATA
class FabricaTransporte {
    criarVeiculoIndividual() { throw new Error("Método abstrato"); }
    criarVeiculoColetivo() { throw new Error("Método abstrato"); }
}

// 4. FABRICAS CONCRETAS
class FabricaTerrestre extends FabricaTransporte {
    criarVeiculoIndividual() { return new Carro(); }
    criarVeiculoColetivo() { return new Onibus(); }
}

class FabricaAerea extends FabricaTransporte {
    criarVeiculoIndividual() { return new Helicoptero(); }
    criarVeiculoColetivo() { return new Aviao(); }
}

// 5. CLIENTE
function iniciarTransporte(fabrica) {
    const individual = fabrica.criarVeiculoIndividual();
    const coletivo = fabrica.criarVeiculoColetivo();

    console.log(" -> " + individual.iniciarRota());
    console.log(" -> " + coletivo.iniciarRota());
    console.log(" ----------------------------");
}

console.log("\n--- AMBIENTE TERRESTRE ---");
iniciarTransporte(new FabricaTerrestre());

console.log("\n--- AMBIENTE AÉREO ---");
iniciarTransporte(new FabricaAerea());

File: Abstract Factory/Exercicios/Ex1_Transporte/Ex1.java
/**
 * PADRÃO: ABSTRACT FACTORY (FÁBRICA ABSTRATA)
 * 
 * Exercício: Sistema de Transporte.
 * O objetivo é criar famílias de veículos (Individual e Coletivo) 
 * que variam de acordo com o ambiente (Terrestre ou Aéreo).
 */

// 1. PRODUTOS ABSTRATOS
abstract class VeiculoIndividual {
    public abstract String iniciarRota();
}

abstract class VeiculoColetivo {
    public abstract String iniciarRota();
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA TERRESTRE
class Carro extends VeiculoIndividual {
    @Override
    public String iniciarRota() {
        return "Partiu! Carro na pista (Transporte Terrestre Individual).";
    }
}

class Onibus extends VeiculoColetivo {
    @Override
    public String iniciarRota() {
        return "Lotou! Ônibus saindo do ponto (Transporte Terrestre Coletivo).";
    }
}

// 2. PRODUTOS CONCRETOS - FAMÍLIA AÉREA
class Helicoptero extends VeiculoIndividual {
    @Override
    public String iniciarRota() {
        return "Hélice girando. Decolagem autorizada! (Transporte Aéreo Individual).";
    }
}

class Aviao extends VeiculoColetivo {
    @Override
    public String iniciarRota() {
        return "Atenção passageiros, iniciando voo (Transporte Aéreo Coletivo).";
    }
}

// 3. FABRICA ABSTRATA
abstract class FabricaTransporte {
    public abstract VeiculoIndividual criarVeiculoIndividual();
    public abstract VeiculoColetivo criarVeiculoColetivo();
}

// 4. FABRICAS CONCRETAS
class FabricaTerrestre extends FabricaTransporte {
    @Override
    public VeiculoIndividual criarVeiculoIndividual() {
        return new Carro();
    }

    @Override
    public VeiculoColetivo criarVeiculoColetivo() {
        return new Onibus();
    }
}

class FabricaAerea extends FabricaTransporte {
    @Override
    public VeiculoIndividual criarVeiculoIndividual() {
        return new Helicoptero();
    }

    @Override
    public VeiculoColetivo criarVeiculoColetivo() {
        return new Aviao();
    }
}

// 5. CLIENTE
public class Ex1 {
    public static void iniciarTransporte(FabricaTransporte fabrica) {
        // O cliente usa a fábrica para criar os produtos compatíveis
        VeiculoIndividual individual = fabrica.criarVeiculoIndividual();
        VeiculoColetivo coletivo = fabrica.criarVeiculoColetivo();

        System.out.println(" -> " + individual.iniciarRota());
        System.out.println(" -> " + coletivo.iniciarRota());
        System.out.println(" ----------------------------");
    }

    public static void main(String[] args) {
        System.out.println("\n--- AMBIENTE TERRESTRE ---");
        iniciarTransporte(new FabricaTerrestre());

        System.out.println("\n--- AMBIENTE AÉREO ---");
        iniciarTransporte(new FabricaAerea());
    }
}